src/validation.cpp

```diff
@@ -4626,6 +4626,22 @@ int VersionBitsTipStateSinceHeight(const Consensus::Params& params, Consensus::D
     return VersionBitsStateSinceHeight(::ChainActive().Tip(), params, pos, versionbitscache);
 }
 
+int64_t VersionBitsActivationHeight(const Consensus::Params& params, Consensus::DeploymentPos pos)
```

Comment:
done

Reply:
Making Period() a public member is cheating :)

Reply:
urgh. _None_ of the BIP9 code is currently used. I've now moved logic from `rpc/blockchain.cpp` into `validation.cpp` after @sdaftuar's comment here: https://github.com/bitcoin/bitcoin/pull/16060#discussion_r305502345, then pulled it into `versionbits.cpp` because of your comment here https://github.com/bitcoin/bitcoin/pull/16060#discussion_r308028983.

I can add even more machinery to versionbits.cpp to avoid making `Period()` public, but I don't understand where we're going with this. Once again, this is code that is _not_ expected to be run. This PR was supposed to be reducing the code complexity, but successive comments have asked for more and more code to be added to cover unexercised code paths.

Reply:
Sorry, that was just meant as a tease, not a criticism or a suggestion to change it. The ThresholdConditionChecker API isn't great.